icu4c/source/i18n/dtptngen.cpp

```diff
@@ -803,38 +803,62 @@ DateTimePatternGenerator::staticGetBaseSkeleton(
-        DateFormat::EStyle style = static_cast<DateFormat::EStyle>(i);
-        df = DateFormat::createDateInstance(style, locale);
-        SimpleDateFormat* sdf;
-        if (df != nullptr && (sdf = dynamic_cast<SimpleDateFormat*>(df)) != nullptr) {
-            sdf->toPattern(dfPattern);
-            addPattern(dfPattern, false, conflictingString, status);
-        }
-        // TODO Maybe we should return an error when the date format isn't simple.
-        delete df;
-        if (U_FAILURE(status)) { return; }
+    if (U_FAILURE(status)) {
+        return;
+    }
+    
+    LocalUResourceBundlePointer rb(ures_open(nullptr, locale.getBaseName(), &status));
+    CharString calendarTypeToUse; // to be filled in with the type to use, if all goes well
+    getCalendarTypeToUse(locale, calendarTypeToUse, status);
 
-        df = DateFormat::createTimeInstance(style, locale);
-        if (df != nullptr && (sdf = dynamic_cast<SimpleDateFormat*>(df)) != nullptr) {
-            sdf->toPattern(dfPattern);
-            addPattern(dfPattern, false, conflictingString, status);
+    // HACK to get around the fact that the old SimpleDateFormat code (actually, Calendar::getCalendarTypeForLocale() )
+    // returns "gregorian" for ja_JP_TRADITIONAL instead of "japanese"
+    if (uprv_strcmp(locale.getBaseName(), "ja_JP_TRADITIONAL") == 0) {
+        calendarTypeToUse.clear().append("gregorian", status);
+    }
+    
+    if (U_FAILURE(status)) {
+        return;
+    }
 
-            // TODO: C++ and Java are inconsistent (see #12568).
-            // C++ uses MEDIUM, but Java uses SHORT.
-            if ( i==DateFormat::kShort && !dfPattern.isEmpty() ) {
-                consumeShortTimePattern(dfPattern, status);
-            }
+    // NOTE: It probably makes more sense to also iterate over the DateTimeSkeletons resource and add the
+    // standard patterns with their corresponding skeletons, but this doesn't currently work because the
+    // CLDR data has too many examples of multiple (different) standard patterns with the same skeleton
+    // (not to mention standard patterns whose skeleton conflicts with an entry in availableFormats).
+    // So we probably can't do anything "smarter" without cleanup at the CLDR level (cleanup that might
+    // just obviate the need for this whole function).  --rtg 2/1/24
+    CharString patternResourcePath;
+    patternResourcePath.append(DT_DateTimeCalendarTag, status)
+        .append('/', status)
+        .append(calendarTypeToUse, status)
+        .append('/', status)
+        .append(DT_DateTimePatternsTag, status);
+
+    LocalUResourceBundlePointer dateTimePatterns;
+    dateTimePatterns.adoptInstead(
+        ures_getByKeyWithFallback(rb.getAlias(), patternResourcePath.data(),
+                                  (UResourceBundle*)nullptr, &status));
+    if (ures_getSize(dateTimePatterns.getAlias()) < 8 || ures_getType(dateTimePatterns.getAlias()) != URES_ARRAY) {
```

Comment:
Shouldn't these two checks be reversed, first verifying that it is the expected type and then verifying the size of it? If it's an unknown type, it's unknown what the size means, isn't it?

Reply:
Yeah, I guess you're right...